ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -425,26 +417,64 @@ impl<'a> Assemble<'a> {
             let guardcheck_label = self.asm.new_dynamic_label();
             for (deoptid, fail_label) in infos {
                 self.comment(
                     self.asm.offset(),
-                    format!("Deopt ID for guard {:?}", deoptid),
+                    format!("Deopt ID and patch point for guard {:?}", deoptid),
                 );
                 self.ra
                     .get_ready_for_deopt(&mut self.asm, &self.deoptinfo[&deoptid].0);
                 // FIXME: Why are `deoptid`s 64 bit? We're not going to have that many guards!
+
+                // Align this location to the cache-line, which since a cache-line is a multiple of
+                // 8 bytes, will also word-align it too. This allows us to safely patch this place
+                // later in a single atomic operation.
+                let clsize =
+                    cache_size::cache_line_size(1, cache_size::CacheType::Instruction).unwrap();
+                // Unwrap can only fail if there's an overflow.
+                self.asm.align(clsize, NOP_OPCODE);
+                // Store the future patch offset for this guard.
+                let off = self.asm.offset();
+                self.deoptinfo.get_mut(&deoptid).unwrap().1.fail_offset = off;
+                // Emit the guard failure code.
+                let jumpreg = self.patch_reg[&deoptid];
                 let deoptid = i32::try_from(deoptid).unwrap();
                 dynasm!(self.asm
                     ;=> fail_label
+                    // After compiling a side-trace for this location, we want to patch in a jump
+                    // to its address here later, so we can jump to side-traces directly.
+                    // We use an available register that the register allocator has reserved for us
+                    // to store the jump location. If the register is 0, the normal deopt routine
+                    // is executed. Once a side-trace is compiled, we patch the mov instruction
+                    // below with a side-trace address to allow us to directly jump to the
+                    // side-trace.
+                    // FIXME: If the side-trace offset to its parent-trace is < 32-bit we can emit
+                    // a relative jump here that doesn't require a register and is likely faster.
+                    // FIXME: Ideally, instead of patching this place, we could patch the guards
+                    // directly which gets rid of an extra jump. But since `cg_icmp_guard` makes
+                    // use of various types of jumps (e.g. jne, jl, etc), this is an optimisation
+                    // for another time.
+                    //
+                    // Since the mov instruction's non-operand part is 2 bytes, push 6 bytes here
+                    // so the address operand is 8-byte aligned.
+                    ; nop; nop; nop; nop; nop; nop
+                    ; mov Rq(jumpreg.code()), QWORD 0x0
```

Comment:
Maybe I'm a dummy but I *though* we were going to do something like:

```
MOV Rq(jumpreg.code()), QWORD deopt
JMP Rq(jumpreg.code())
```

without the `cmp/je` dance?

I thought that the problem you had with this is that global labels don't work with `MOV` or similar in dynasm? [Did we confirm this?] If so, we could patch this address before codegen finishes. [It would be nicer to use labels, but maybe that doesn't work.]

Reply:
Yeah, we can't use labels as they emit a different mov instruction that we can't patch later with an absolute address. So I will do the patching.

Reply:
Maybe I'm wrong but isn't it something like:

```rust
let cur_addr = asm.len_bytes();
let mut align = (cur_addr+2).next_multiple_of(8);
if align.next_multiple_of(page_size) == align { align += 8 }
let num_nops = align - cur_addr;
```

?

Reply:
[10b0e34](https://github.com/ykjit/yk/pull/1591/commits/10b0e34a22b294dac02b12ec6959bb8005e195ec)